pkg/placementrule/controller/placementrule/placementdecision.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2021 The Kubernetes Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package placementrule
+
+import (
+	"context"
+	"fmt"
+
+	errorhelpers "github.com/openshift/library-go/pkg/operator/v1helpers"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog"
+	clusterapi "open-cluster-management.io/api/cluster/v1alpha1"
+	placementruleapi "open-cluster-management.io/multicloud-operators-subscription/pkg/apis/apps/placementrule/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	placementRuleLabel       = "cluster.open-cluster-management.io/placementrule"
+	maxNumOfClusterDecisions = 100
+)
+
+// syncPlacementDecisions create/update/delete placementdecisions based on PlacementRule's status.decisions
+// based on https://github.com/open-cluster-management-io/placement/blob/v0.2.0/pkg/controllers/scheduling/scheduling_controller.go#L339
+func (r *ReconcilePlacementRule) syncPlacementDecisions(ctx context.Context,
+	prNamespace string, prName string) error {
+	klog.Info("syncPlacementDecisions placementrule ", prNamespace, "/", prName)
+
+	prNsn := types.NamespacedName{Namespace: prNamespace, Name: prName}
+
+	placementRule := &placementruleapi.PlacementRule{}
+
+	err := r.Get(ctx, prNsn, placementRule)
+	if err != nil {
+		return err
+	}
+
+	var prDecisions []placementruleapi.PlacementDecision
+
+	if placementRule.Status.Decisions != nil && len(placementRule.Status.Decisions) > 0 {
+		prDecisions = placementRule.Status.Decisions
+	}
+
+	// no sorting necessary because it's already sorted by placementRule
+
+	var clusterDecisions []clusterapi.ClusterDecision
+
+	for _, prDecision := range prDecisions {
+		clusterDecisions = append(clusterDecisions, clusterapi.ClusterDecision{ClusterName: prDecision.ClusterName})
+	}
+
+	// split the cluster decisions into slices, the size of each slice cannot exceed
+	// maxNumOfClusterDecisions.
```

Comment:
The `maxNumOfClusterDecisions` is hard set to 100. So each placementDecision CR can contain no more than 100 clusters.  

Based on the pervious placementRule test on 2k managed clusters, one placementRule CR status can contain 2k clusters without any performance issue.

So I think we can simply just create one placementDecision CR for holding all clusters from its associated placementRule status. (1 placementDecsion mapping to 1 placementRule  )  This will simplify the implementation a lot.

There is another side effect if there are multiple placementDecsions mapping to 1 placementRule, creating/updating the placementDecision CR  just based on the placementDecision name may not be engough. As there is no guarantee for GO for loop to get the exact result each time, the placmentDecison CR with the same name may have different clusters.

Also watching multiple placementDecision in subscription controller will be more cafeful. - When a placementDecision is reconciled, the subscription controller have to get all other placementDecision CRs to get the full cluster list.